# 데이터 변환 (p.54)
- pivot을 이용하여 GENDER와 AGEGROUP별 WEIGHT의 평균을 교재와 같이 변환하라.

## 1. 데이터 구성하기

In [1]:
import pandas as pd

fit=pd.read_csv("C:/Users/Myung-jin/Git_Temp-main/FITNESS.CSV")
fit

NAME GENDER  AGE  AGEGROUP  WEIGHT     OXY  RUNTIME  RUNPULSE  \
0      Donna      F   42        40   68.15  59.571     8.17       166   
1     Gracie          38        30   81.87  60.055     8.63       170   
2     Luanne      F   43        40   85.84  54.297     8.65       156   
3       Mimi      F   50        50   70.87  54.625     8.92       146   
4      Chris      M   49        40   70.00  49.156     8.95       180   
5      Allen      M   38        30   89.02  49.874     9.22       178   
6      Nancy      F   49        40   76.32  48.673     9.40       186   
7      Patty      F   52        50   76.32  45.441     9.63       164   
8    Suzanne      F   57        50   59.08  50.545     9.93       148   
9     Teresa      F   51        50   77.91  46.672    10.00       162   
10       Bob      M   39        30   75.07  45.313    10.07       185   
11  Harriett      F   49        40   73.37  50.388    10.08       168   
12      Jane      F   44        40   73.03  50.541    10.13       168   
13    Harold      M   48        40   91.63  46.774    10.25       162   
14     Sammy      M   54        50   83.12  51.855    10.33       166   
15     Buffy      F   52        50   73.71  45.790    10.47       186   
16     Trent      M   38        30   82.78  47.467    10.50       170   
17    Jackie      F   47        40   79.15  47.273    10.60       162   
18     Ralph      M   43        40   66.00  49.091    10.85       162   
19      Jack      M   51        50   69.63  40.836    10.95       168   
20     Annie      F   51        50   67.25  45.118    11.08       172   
21      Kate      F   45        40   66.45  44.754    11.12       176   
22      Carl      M   54        50   79.38  46.080    11.17       156   
23       Don      M   44        30   89.47  44.609    11.37       178   
24     Effie      F   48        40   61.24  47.920    11.50       170   
25    George      M   37        30   77.45  44.811    11.63       176   
26      Iris      F   39        30   75.98  45.681    11.95       176   
27      Mark      M   57        50   73.37  39.407    12.63       174   
28     Steve      M   54        50   91.63  39.203    12.88       168   
29    Vaughn      M   44        40   81.42  39.442    13.08       174   
30   William      M   45        40   87.66  37.388    14.03       186   

    RSTPULSE  MAXPULSE  
0         40       172  
1         48       186  
2         45       168  
3         48       155  
4         44       185  
5         55       180  
6         56       188  
7         48       166  
8         49       155  
9         48       168  
10        62       185  
11        67       168  
12        45       168  
13        48       164  
14        50       170  
15        59       188  
16        53       172  
17        47       164  
18        64       170  
19        57       172  
20        48       172  
21        51       176  
22        62       165  
23        62       182  
24        52       176  
25        58       176  
26        70       180  
27        58       176  
28       144       172  
29        63       176  
30        56       192

## 1-1 데이터 가공하기

In [2]:
# GENDER 그룹의 2번째 행에 값 채우기
fit.iloc[1, 1] = 'F'
fit.head()

NAME GENDER  AGE  AGEGROUP  WEIGHT     OXY  RUNTIME  RUNPULSE  RSTPULSE  \
0   Donna      F   42        40   68.15  59.571     8.17       166        40   
1  Gracie      F   38        30   81.87  60.055     8.63       170        48   
2  Luanne      F   43        40   85.84  54.297     8.65       156        45   
3    Mimi      F   50        50   70.87  54.625     8.92       146        48   
4   Chris      M   49        40   70.00  49.156     8.95       180        44   

   MAXPULSE  
0       172  
1       186  
2       168  
3       155  
4       185

In [3]:
# 잘못 설정된 AGEGROUP 수정해주기
fit.iloc[:, 3] = fit.iloc[:, 2] - (fit.iloc[:,2]%10)
fit

NAME GENDER  AGE  AGEGROUP  WEIGHT     OXY  RUNTIME  RUNPULSE  \
0      Donna      F   42        40   68.15  59.571     8.17       166   
1     Gracie      F   38        30   81.87  60.055     8.63       170   
2     Luanne      F   43        40   85.84  54.297     8.65       156   
3       Mimi      F   50        50   70.87  54.625     8.92       146   
4      Chris      M   49        40   70.00  49.156     8.95       180   
5      Allen      M   38        30   89.02  49.874     9.22       178   
6      Nancy      F   49        40   76.32  48.673     9.40       186   
7      Patty      F   52        50   76.32  45.441     9.63       164   
8    Suzanne      F   57        50   59.08  50.545     9.93       148   
9     Teresa      F   51        50   77.91  46.672    10.00       162   
10       Bob      M   39        30   75.07  45.313    10.07       185   
11  Harriett      F   49        40   73.37  50.388    10.08       168   
12      Jane      F   44        40   73.03  50.541    10.13       168   
13    Harold      M   48        40   91.63  46.774    10.25       162   
14     Sammy      M   54        50   83.12  51.855    10.33       166   
15     Buffy      F   52        50   73.71  45.790    10.47       186   
16     Trent      M   38        30   82.78  47.467    10.50       170   
17    Jackie      F   47        40   79.15  47.273    10.60       162   
18     Ralph      M   43        40   66.00  49.091    10.85       162   
19      Jack      M   51        50   69.63  40.836    10.95       168   
20     Annie      F   51        50   67.25  45.118    11.08       172   
21      Kate      F   45        40   66.45  44.754    11.12       176   
22      Carl      M   54        50   79.38  46.080    11.17       156   
23       Don      M   44        40   89.47  44.609    11.37       178   
24     Effie      F   48        40   61.24  47.920    11.50       170   
25    George      M   37        30   77.45  44.811    11.63       176   
26      Iris      F   39        30   75.98  45.681    11.95       176   
27      Mark      M   57        50   73.37  39.407    12.63       174   
28     Steve      M   54        50   91.63  39.203    12.88       168   
29    Vaughn      M   44        40   81.42  39.442    13.08       174   
30   William      M   45        40   87.66  37.388    14.03       186   

    RSTPULSE  MAXPULSE  
0         40       172  
1         48       186  
2         45       168  
3         48       155  
4         44       185  
5         55       180  
6         56       188  
7         48       166  
8         49       155  
9         48       168  
10        62       185  
11        67       168  
12        45       168  
13        48       164  
14        50       170  
15        59       188  
16        53       172  
17        47       164  
18        64       170  
19        57       172  
20        48       172  
21        51       176  
22        62       165  
23        62       182  
24        52       176  
25        58       176  
26        70       180  
27        58       176  
28       144       172  
29        63       176  
30        56       192

In [4]:
# 특정열만 불러오기
fit1 = fit[['GENDER', 'AGEGROUP', 'WEIGHT']]
fit1

GENDER  AGEGROUP  WEIGHT
0       F        40   68.15
1       F        30   81.87
2       F        40   85.84
3       F        50   70.87
4       M        40   70.00
5       M        30   89.02
6       F        40   76.32
7       F        50   76.32
8       F        50   59.08
9       F        50   77.91
10      M        30   75.07
11      F        40   73.37
12      F        40   73.03
13      M        40   91.63
14      M        50   83.12
15      F        50   73.71
16      M        30   82.78
17      F        40   79.15
18      M        40   66.00
19      M        50   69.63
20      F        50   67.25
21      F        40   66.45
22      M        50   79.38
23      M        40   89.47
24      F        40   61.24
25      M        30   77.45
26      F        30   75.98
27      M        50   73.37
28      M        50   91.63
29      M        40   81.42
30      M        40   87.66

## 2. groupby, agg를 이용한 평균값 계산

In [5]:
df_group = fit1.groupby(['GENDER', 'AGEGROUP']).agg('mean')
df_group

WEIGHT
GENDER AGEGROUP           
F      30        78.925000
       40        72.943750
       50        70.856667
M      30        81.080000
       40        81.030000
       50        79.426000

In [6]:
df_group.reset_index(inplace = True)
df_group

GENDER  AGEGROUP     WEIGHT
0      F        30  78.925000
1      F        40  72.943750
2      F        50  70.856667
3      M        30  81.080000
4      M        40  81.030000
5      M        50  79.426000

In [7]:
# 데이터값 수정하기
df_group.replace('F', '여성', inplace = True)
df_group.replace('M', '남성', inplace = True)
df_group

GENDER  AGEGROUP     WEIGHT
0     여성        30  78.925000
1     여성        40  72.943750
2     여성        50  70.856667
3     남성        30  81.080000
4     남성        40  81.030000
5     남성        50  79.426000

In [8]:
df_group.replace({'AGEGROUP' : {30: '30대'}}, inplace = True)
df_group.replace({'AGEGROUP' : {40: '40대'}}, inplace = True)
df_group.replace({'AGEGROUP' : {50: '50대'}}, inplace = True)
df_group

GENDER AGEGROUP     WEIGHT
0     여성      30대  78.925000
1     여성      40대  72.943750
2     여성      50대  70.856667
3     남성      30대  81.080000
4     남성      40대  81.030000
5     남성      50대  79.426000

In [9]:
df_group.iloc[:, 2] = df_group.iloc[:, 2].round(3)
df_group

GENDER AGEGROUP  WEIGHT
0     여성      30대  78.925
1     여성      40대  72.944
2     여성      50대  70.857
3     남성      30대  81.080
4     남성      40대  81.030
5     남성      50대  79.426

## 3. pivot을 이용한 변환

In [10]:
# pivot을 이용하여 index는 GENDER, column은 AGEGROUP으로 설정하여 변환
df_group = df_group.pivot(index = 'GENDER', columns = 'AGEGROUP', values = 'WEIGHT')
df_group

AGEGROUP     30대     40대     50대
GENDER                          
남성        81.080  81.030  79.426
여성        78.925  72.944  70.857

In [11]:
# 행, 열 변경
df_group.T

GENDER        남성      여성
AGEGROUP                
30대       81.080  78.925
40대       81.030  72.944
50대       79.426  70.857

**=> 교재와 같이 데이터 변환을 완료하였으나, 40대-남성의 weight 평균이 일치하지 않았다. 
원래 데이터에서 해당하는 수치의 mean값을 구해보았더니 81.030이 맞았다. 교재와 수치가 완전하게 일치하진 않지만 문제의 원인을 해결할 수 없다고 결론을 냈고, 데이터 수치나 교재 자료의 오류라고 판단한다.**